synapse/lib/auth.py

```diff
@@ -81,3 +97,900 @@ def runas(user):
+                                    mesg='node:del requires "form"')
+        self._node_del[form] = True
+
+    def _addNodeSet(self, rule):
+        form = rule[1].get('form')
+        prop = rule[1].get('prop')
+        if not form:
+            raise s_exc.BadRuleValu(key='form', valu=form,
+                                    mesg='node:set:prop requires "form"')
+        if not prop:
+            raise s_exc.BadRuleValu(key='valu', valu=prop,
+                                    mesg='node:set:prop requires "prop"')
+        self._node_set[(form, prop)] = True
+
+    def _addTagAdd(self, rule):
+        tag = rule[1].get('tag')
+        if not tag:
+            raise s_exc.BadRuleValu(key='tag', valu=tag,
+                                    mesg='node:tag:add requires "tag"')
+        self._tag_add.add(tag)
+
+    def _addTagDel(self, rule):
+        tag = rule[1].get('tag')
+        if not tag:
+            raise s_exc.BadRuleValu(key='tag', valu=tag,
+                                    mesg='node:tag:del requires "tag"')
+        self._tag_del.add(tag)
+
+    #####################################################
+
+    def _mayNodeAdd(self, perm):
+
+        form = perm[1].get('form')
+
+        if self._node_add.get(form):
+            return True
+
+        if self._node_add.get('*'):
+            return True
+
+        return False
+
+    def _mayNodeDel(self, perm):
+
+        form = perm[1].get('form')
+
+        if self._node_del.get(form):
+            return True
+
+        if self._node_del.get('*'):
+            return True
+
+        return False
+
+    def _mayNodeSet(self, perm):
+
+        form = perm[1].get('form')
+        prop = perm[1].get('prop')
+
+        if self._node_set.get((form, prop)):
+            return True
+
+        if self._node_set.get((form, '*')):
+            return True
+
+        if self._node_set.get(('*', '*')):
+            return True
+
+        return False
+
+    def _mayTagAdd(self, perm):
+        tag = perm[1].get('tag')
+        return self._tag_add.get(tag)
+
+    def _mayTagDel(self, perm):
+        tag = perm[1].get('tag')
+        return self._tag_del.get(tag)
+
+    #####################################################
+
+class Role(AuthBase):
+
+    def _saveAuthData(self):
+        info = {
+            'admin': self.admin,
+            'rules': self.rules
+        }
+        self.auth._saveRoleInfo(self.name, info)
+
+class User(AuthBase):
+
+    def _getAuthData(self):
+        info = AuthBase._getAuthData(self)
+        info['roles'] = list(self.roles.keys())
+        return info
+
+    def _saveAuthData(self):
+        info = self._getAuthData()
+        self.auth._saveUserInfo(self.name, info)
+
+    def _initAuthData(self):
+
+        AuthBase._initAuthData(self)
+
+        self.roles = {}
+
+        for name in self.info.get('roles', ()):
+
+            role = self.auth.roles.get(name)
+
+            if role is None:  # pragma: no cover
```

Comment:
Is it possible to test this section?

Reply:
We'd have to hand-jam a user struct into the lmdb instance directly to test it.